File: public/templates/fenwick_tree.py
class BIT:
    def __init__(self,n:int):
       self.nums = [0]*(n+1)
    def update(self,i:int,val:int): # update rank i by val
        i +=1
        while(i<len(self.nums)):
            self.nums[i] +=val
            i += (i&(-i))
    def sum(self,i:int) -> int: # sum of rank 0...i-1
        r = 0
        # i +=1, not needed here since we need sum of rank less than i rank[0...i-1]
        while(i>0):
            r += self.nums[i]
            i -= (i&(-i))
        return r
    

# class MaxBIT:  # One-based indexing
#     def __init__(self, size):
#         self.bit = [0] * (size + 1)
#     def get(self, idx):
#         ans = 0
#         while idx > 0:
#             ans = max(ans, self.bit[idx])
#             idx -= idx & (-idx)
#         return ans
#     def update(self, idx, val):
#         while idx < len(self.bit):
#             self.bit[idx] = max(self.bit[idx], val)
#             idx += idx & (-idx)

File: public/templates/fenwick_tree.cpp
#include<bits/stdc++.h>
#include <vector>
typedef long long ll;
class BIT {
private:
    std::vector<int> nums;
    int LOG;
public:
    BIT(int n) {
    	LOG = (int)log2(n)+1;
        nums.resize((1ll<<LOG) + 1);
    }
    void update(int i, int val) {
        i += 1;
        while (i < nums.size()) {
            nums[i] += val;
            i += (i & (-i));
        }
    }
    int sum(int i) {
        int r = 0;
        // i += 1, not needed here since we need sum of rank less than i rank[0...i-1]
        while (i > 0) {
            r += nums[i];
            i -= (i & (-i));
        }
        return r;
    }
    // max index where sum A[0...ind] < x,return index of first number greator than or equal to x
    int lower(int x){
        ll pref = 0;
        int ind = 0;
        for(int i=LOG;i>=0;i--){
            if(nums[ind + (1ll<<i)] + pref<x){
                pref += nums[ind + (1ll<<i)];
                ind += 1ll<<i;
            }
        }
        return ind; // 0  based
    }
    
};

#include <cassert>

int main() {
    BIT bit(5);
    bit.update(0, 1);
    bit.update(1, 2);
    bit.update(2, 3);
    bit.update(3, 4);
    bit.update(4, 5);
    assert(bit.sum(0) == 0);
    assert(bit.sum(1) == 1);
    assert(bit.sum(2) == 3);
    assert(bit.sum(3) == 6);
    assert(bit.sum(4) == 10);
    assert(bit.sum(5) == 15);
    assert(bit.lower(1) == 1);
    assert(bit.lower(2) == 2);
    assert(bit.lower(3) == 2);
    return 0;
}